internal/app/machined/pkg/controllers/config/acquire.go

```diff
@@ -0,0 +1,437 @@
+	if ctrl.ConfigPath == "" {
+		ctrl.ConfigPath = constants.ConfigPath
+	}
+
+	// start always with loading config from disk
+	var currentState stateMachineFunc = ctrl.stateDisk
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		// check the spec first
+		_, err := safe.ReaderGet[*v1alpha1.AcquireConfigSpec](ctx, r, v1alpha1.NewAcquireConfigSpec().Metadata())
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// spec is not found, wait for it
+				continue
+			}
+
+			return fmt.Errorf("failed to get spec: %w", err)
+		}
+
+		// run the state machine
+		for {
+			newState, cfg, err := currentState(ctx, r, logger)
+			if err != nil {
+				ctrl.EventPublisher.Publish(ctx, &machineapi.ConfigLoadErrorEvent{
+					Error: err.Error(),
+				})
+
+				ctrl.PlatformEvent.FireEvent(
+					ctx,
+					platform.Event{
+						Type:    platform.EventTypeFailure,
+						Message: "Error loading and validating Talos machine config.",
+						Error:   err,
+					},
+				)
+
+				return err
+			}
+
+			if cfg != nil {
+				// apply config
+				if err = ctrl.ConfigSetter.SetConfig(cfg); err != nil {
+					return fmt.Errorf("failed to set config: %w", err)
+				}
+			}
+
+			if newState == nil {
+				// wait for reconcile event, keep running in the same state
+				break
+			}
+
+			currentState = newState
+		}
+
+		r.ResetRestartBackoff()
+	}
+}
+
+// stateDisk acquires machine configuration from disk (STATE partition).
+//
+// Transitions:
+//
+//	--> platform: no config found on disk, proceed to platform
+//	--> maintenanceEnter: config found on disk, but it's incomplete, proceed to maintenance
+//	--> done: config found on disk, and it's complete
+func (ctrl *AcquireController) stateDisk(ctx context.Context, r controller.Runtime, logger *zap.Logger) (stateMachineFunc, config.Provider, error) {
+	cfg, err := ctrl.loadFromDisk(logger)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	switch {
+	case cfg == nil:
+		// no config loaded, proceed to platform
+		return ctrl.statePlatform, nil, nil
+	case cfg.CompleteForBoot():
+		// complete config, we are done
+		return ctrl.stateDone, cfg, nil
+	default:
+		// incomplete config, proceed to maintenance
+		return ctrl.stateMaintenanceEnter, cfg, nil
+	}
+}
+
+// loadFromDisk is a helper function for stateDisk.
+func (ctrl *AcquireController) loadFromDisk(logger *zap.Logger) (config.Provider, error) {
+	logger.Debug("loading config from STATE", zap.String("path", ctrl.ConfigPath))
+
+	_, err := os.Stat(ctrl.ConfigPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// no saved machine config
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("failed to stat %s: %w", ctrl.ConfigPath, err)
+	}
+
+	cfg, err := configloader.NewFromFile(ctrl.ConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load config from STATE: %w", err)
+	}
+
+	// this is legacy flow with persistence, it should be deprecated & removed
```

Comment:
also add a TODO maybe, so it's easier to search

Reply:
done